src/auth.zig

```diff
@@ -107,6 +97,320 @@ test MTLS {
+            const ca: *Self = @ptrCast(@alignCast(ptr));
+            if (ca.base) |base| return base.lookupUser(user_id);
+            return error.UnknownUser;
+        }
+
+        pub fn mkToken(hm: *HMAC, token: []u8, user: *const User) Error!usize {
+            var buffer: [Self.ibuf_size]u8 = [_]u8{0} ** Self.ibuf_size;
+            buffer[0] = Token.Version;
+            var b: []u8 = buffer[1..];
+            const time = toBytes(nativeToLittle(i64, std.time.timestamp()));
+            hm.update(time[0..8]);
+            @memcpy(b[0..8], time[0..8]);
+            b = b[8..];
+
+            if (user.username) |un| {
+                if (un.len > b.len - HMAC.mac_length - 1) return error.NoSpaceLeft;
+                hm.update(un);
+                @memcpy(b[0..un.len], un);
+                b[un.len] = ':';
+                b = b[un.len + 1 ..];
+            }
+
+            if (user.session_extra_data) |ed| {
+                if (ed.len > b.len - HMAC.mac_length - 1) return error.NoSpaceLeft;
+                hm.update(ed);
+                @memcpy(b[0..ed.len], ed);
+                b[ed.len] = ':';
+                b = b[ed.len + 1 ..];
+            }
+            hm.final(b[0..HMAC.mac_length]);
+            b = b[HMAC.mac_length..];
+
+            const final = buffer[0 .. buffer.len - b.len];
+            if (token.len < b64_enc.calcSize(final.len)) return error.NoSpaceLeft;
+            return b64_enc.encode(token, final).len;
+        }
+
+        pub fn createSession(ptr: *anyopaque, user: *User) Error!void {
+            const ca: *Self = @ptrCast(@alignCast(ptr));
+            if (ca.base) |base| base.createSession(user) catch |e| switch (e) {
+                error.NotProvided => {},
+                else => return e,
+            };
+
+            const prefix_len: usize = (if (user.username) |u| u.len + 1 else 0) +
+                if (user.session_extra_data) |ed| ed.len + 1 else 0;
+            if (prefix_len > ca.session_buffer.len / 2) {
+                if (ca.alloc == null) return error.NoSpaceLeft;
+                return error.NoSpaceLeft;
+            }
+
+            var hmac = HMAC.init(ca.server_secret_key);
+            const len = try mkToken(&hmac, ca.session_buffer[0..], user);
+            user.session_next = ca.session_buffer[0..len];
+        }
+
+        pub fn getCookie(ptr: *anyopaque, user: User) Error!?ReqCookie {
+            const ca: *Self = @ptrCast(@alignCast(ptr));
+            if (ca.base) |base| if (base.vtable.get_cookie) |_| {
+                return base.getCookie(user);
+            };
+
+            if (user.session_next) |next| {
+                return .{
+                    .name = "verse_session_secret",
+                    .value = next,
+                    .attr = .{
+                        // TODO only set when HTTPS is enabled
+                        .secure = true,
+                        .same_site = .strict,
```

Comment:
The expires attribute should also be set.

Reply:
I didn't implement it because expires requires a timestamp; I can set max-age if you think it's important? Otherwise, I'd default to keeping the cookie around, and making a decision based on it, even if it's expired. E.g. 2fa you may have an expired cookie, but it's still otherwise valid. You could just re-auth with 2fa instead of a strict-reauth

Reply:
I'm gonna double down on this, I don't wanna send the max age to the client, because that's a disclosure that doesn't need to happen.

Reply:
I think your reasons for not setting it make sense.